README
ksdjfksjdfkdsf
$Id$


INTRODUCTION

  OpenDNSSEC was created as an open-source turn-key solution for DNSSEC.
  It secures zone data just before it is published in an authoritative
  name server.


MORE INFORMATION

  More informnation can be found at the project website available at
  http://www.opendnssec.org/ and on the development WIKI at
  http://wiki.opendnssec.org/.

  Information about announcements, bug reporting and mailing lists can
  be found at http://www.opendnssec.org/support/.


DEPENDENCIES

  OpenDNSSEC depends on  a number of external packages:

  - libmxl2 (including xmllint)
  - LDNS
  - SQLite3

  To run OpenDNSSEC, one must have at least one crypto module providing a
  PKCS#11 library, e.g. SoftHSM (http://www.opendnssec.org/softHSM)

  The Auditor has some additional dependencies:

  - Ruby
  - rubygems
    * syslog
    * openssl
    * dnsruby
    * xsd/datatypes
    * rexml/document

  When building from the subversion repository, the following dependencies
  are also needed:

  - A Java runtime environment (JRE/JDK)



NEWS
kjsdhfkjdshfjdsf
$Id: NEWS 7522 2013-12-18 12:59:18Z matthijs $

OpenDNSSEC 1.4.4rc1

Bugfixes:
* OPENDNSSEC-481: libhsm: Fix an off-by-one length check error.
* OPENDNSSEC-482: libhsm: Improved cleanup for C_FindObjects.
* OPENDNSSEC-515: Signer Engine: Don't replace tabs in RRs with whitespace.
* SUPPORT-86: Fixed build on OS X [OPENDNSSEC-512]

OpenDNSSEC 1.4.3 - 2013-12-04

* SUPPORT-72: Improve logging when failed to increment serial in case 
  of key rollover and serial value "keep" [OPENDNSSEC-461].
* OPENDNSSEC-106: Add 'ods-enforcerd -p <policy>' option. This prompts the
  enforcer to run once and only process the specified policy and associated
  zones.
* OPENDNSSEC-330: NSEC3PARAM TTL can now be optionally configured in kasp.xml. 
  Default value remains PT0S.
* OPENDNSSEC-390: ods-ksmutil: Add an option to the 'ods-ksmutil key ds-seen'
  command so the user can choose not to notify the enforcer.
* OPENDNSSEC-430: ods-ksmutil: Improve 'zone add' - Zone add command 
  could warn if a specified zone file or adapter file does not exits.
* OPENDNSSEC-431: ods-ksmutil: Improve 'zone add' - Support default <input>
  and <output> values for DNS adapters.
* OPENDNSSEC-454: ods-ksmutil: Add option for 'ods-ksmutil key import' to
  check if there is a matching key in the repository before import.

Bugfixes:
* OPENDNSSEC-435: Signer Engine: Fix a serious memory leak in signature
  cleanup.
* OPENDNSSEC-463: Signer Engine: Duration PT0S is now printed correctly.
* OPENDNSSEC-466: Signer Engine: Created bad TSIG signature when falling back
  to AXFR.
* OPENDNSSEC-467: Signer Engine: After ods-signer clear, signer should not use
  inbound serial.


OpenDNSSEC 1.4.2 - 2013-09-11

* OPENDNSSEC-428: ods-ksmutil: Add option for 'ods-ksmutil key generate' to 
  take number of zones as a parameter

Bugfixes:
* SUPPORT-66: Signer Engine: Fix file descriptor leak in case of TCP write
  error [OPENDNSSEC-427].
* SUPPORT-71: Signer Engine: Fix double free crash in case of HSM connection
  error during signing [OPENDNSSEC-444].
* OPENDNSSEC-401: 'ods-signer sign <zone> --serial <nr>' command produces seg
  fault when run directly on command line (i.e. not via interactive mode)
* OPENDNSSEC-440: 'ods-ksmutil key generate' and the enforcer can create
  too many keys if there are keys already available and the KSK and ZSK use
  same algorithm and length
* OPENDNSSEC-424: Signer Engine: Respond to SOA queries from file instead
  of memory. Makes response non-blocking.
* OPENDNSSEC-425 Change "hsmutil list" output so that the table header goes
  to stdout not stderr
* OPENDNSSEC-438: 'ods-ksmutil key generate' and the enforcer can create
  too many keys for <SharedKeys/> policies when KSK and ZSK use same
  algorithm and length
* OPENDNSSEC-443: ods-ksmutil: Clean up of hsm connection handling
* Signer Engine: Improved Inbound XFR checking.
* Signer Engine: Fix double free corruption in case of adding zone with
  DNS Outbound Adapters and NotifyCommand enabled.


OpenDNSSEC 1.4.1 - 2013-06-27

* SUPPORT-58: Extend ods-signer sign <zone> with --serial <nr> so that the user
  can specify the SOA serial to use in the signed zone [OPENDNSSEC-401].
* OPENDNSSEC-91: Make the keytype flag required when rolling keys

Bugfixes:
* SUPPORT-60: Fix datecounter in case inbound serial is higher than outbound
  serial [OPENDNSSEC-420].
* OPENDNSSEC-247: Signer Engine: TTL on NSEC3 was not updated on SOA
  Minimum change.
* OPENDNSSEC-421: Signer Engine: Fix assertion error in case NSEC3 hash
  algorithm in signconf is not SHA1.
* OPENDNSSEC-421: ods-kaspcheck: Check whether NSEC3 hash algorithm in kasp
  is valid.
* Bugfix: The time when inbound serial is acquired was reset invalidly,
  could cause OpenDNSSEC wanting AXFR responses while requesting IXFR (thanks
  Stuart Lau).
* Bugfix: Fix malform in Outbound IXFR/TCP subsequent packet (thanks Stuart
  Lau).
* OPENDNSSEC-398: The ods-ksmutil key rollover command does not work correctly
  when rolling all keys using the --policy option


OpenDNSSEC 1.4.0 - 2013-04-22

* Production release of 1.4
* Versioning scheme and release support policies updated
* Summary of changes in 1.4 can be found on the wiki:
  http://wiki.opendnssec.org/display/DOCS


OpenDNSSEC 1.4.0rc3 - 2013-03-15

* Further testing of OPENDNSSEC-387 completed, release returned to rc status.


OpenDNSSEC 1.4.0b3 - 2013-02-20

Note: This release is marked as a beta release (rather than rc3) due to 
OPENDNSSEC-387, which is a significant functional change compared to rc2. 

* OPENDNSSEC-387: Rollback of multi-threaded enforcer. Due to key allocation 
  issues the usefulness of the threaded enforcer is outweighed by the code
  complications. The option still remains in conf.xml for compatibility with
  existing use; but it will now be silently ignored.

Bugfixes:
* OPENDNSSEC-388: Signer Engine: Internal serial should take into account
  the inbound serial.
* SUPPORT-50/51: Signer Engine: Inbound DNS Adapter incorrectly updates 
  NSEC3PARAM and DNSKEY RRset [OPENDNSSEC-389]
* OPENDNSSEC-389: Input DNS Adapter incorrectly updating NSEC3PARAM and DNSKEY RRsets


OpenDNSSEC 1.4.0rc2 - 2013-01-25

* OPENDNSSEC-350: Signer Engine: Better log message when IXFR is not ready for
  reading.
* OPENDNSSEC-367: ods-ksmutil: Require user confirmation if the algorithm for
  a key is changed in a policy (as this rollover is not handled cleanly)

Bugfixes:
* SUPPORT-44: Signer Engine: Drop privileges after binding to socket
  [OPENDNSSEC-364].
* Signer Engine: XFR not ready should not be a fatal status for task read
  (thanks Ville Mattila).
* OPENDNSSEC-365: Enforcer: Nasty bug where KSKs could get prematurely retired.


OpenDNSSEC 1.4.0rc1 - 2013-01-10

* OPENDNSSEC-359: Remove eppclient


OpenDNSSEC 1.4.0b2 - 2012-12-17

* OPENDNSSEC-292: Provide scripts to convert database between different 
  supported formats
* OPENDNSSEC-299: ods-ksmutil: ods-ksmutil <enter> now includes policy import

home/bjst/parser.c
/*
 * $Id$
 *
 * Copyright (c) 2010 .SE (The Internet Infrastructure Foundation).
 * All rights reserved.
 *
 * Written by Björn Stenberg <bjorn@haxx.se> for .SE
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
 * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
 * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <ctype.h>
#include <unistd.h>
#include <arpa/inet.h>

/* Internal binary RR format:

   ttl      - (int) ttl value.
   owner    - owner name, stored backwards (i.e. 'com.example.www').
              each character is stored in a (short), with these specials:
               -2 = name segment delimiter ('.')
               -1 = end of name
   type     - (unsigned short) RR type
   class    - (unsigned short) RR class
   rdata    - binary RDATA in wire format

   The purpose of using this custom format instead of the exact wire format is
   to enable fast comparison. This way, RR records can be compared using a
   single memcmp() starting at the owner field.
*/

#define MAX_LINE_LEN 65535
#define MAX_NAME_LEN 255

#define NUM_TYPES 101
#define NUM_CLASSES 5

static const char* typename[NUM_TYPES] = {
    NULL, "A", "NS", "MD", "MF", "CNAME", "SOA", "MB", "MG", "MR",
    "NULL","WKS","PTR","HINFO","MINFO","MX","TXT","RP","AFSDB","X25",
    "ISDN","RT","NSAP","NSAP-PTR","SIG","KEY","PX","GPOS","AAAA","LOC",
    "NXT","EID","NIMLOC","SRV","ATMA","NAPTR","KX","CERT","A6","DNAME",
    "SINK","OPT","APL","DS","SSHFP","IPSECKEY","RRSIG","NSEC","DNSKEY","DHCID",
    "NSEC3","NSEC3PARAM",NULL,NULL,NULL,"HIP","NINFO","RKEY",NULL,NULL,
    NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,
    NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,
    NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,
    NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,"SPF","DLV"
};

static const char* classname[NUM_CLASSES] = {
    NULL, "IN", NULL, "CH", "HS"
};

enum {
    RD_INT8,
    RD_INT16,
    RD_INT32,
    RD_NAME,   /* wire format, one string per segment */
    RD_STRING, /* wire format */
    RD_A,
    RD_AAAA,
    RD_LOC,    /* TODO: RFC 1876 */
    RD_BASE64, /* TODO: RFC 4398 (including whitespace!) */
    RD_BASE16, /* TODO: RFC 3658 (hex) */
};

static const char format_list[][8] = {
    /* 0 */         { 0 },
    /* 1: A  */     { 1, RD_A },
    /* 2: NS */     { 1, RD_NAME },
    /* 3: MD */     { 1, RD_NAME },
    /* 4: MF */     { 1, RD_NAME },
    /* 5: CNAME */  { 1, RD_NAME },
    /* 6: SOA */    { 7, RD_NAME, RD_NAME,
                      RD_INT32, RD_INT32, RD_INT32, RD_INT32, RD_INT32 },
    /* 7: MB */     { 1, RD_NAME },
    /* 8: MG */     { 1, RD_NAME },
    /* 9: MR */     { 1, RD_NAME },
    /* 10: NULL */  { 0 }, /* not supported by OpenDNSSEC */
    /* 11: WKS */   { 0 }, /* not supported by OpenDNSSEC */
    /* 12: PTR */   { 1, RD_NAME },
    /* 13: HINFO */ { 2, RD_STRING, RD_STRING },
    /* 14: MINFO */ { 2, RD_NAME, RD_NAME },
    /* 15: MX */    { 2, RD_INT16, RD_NAME },
    /* 16: TXT */   { 1, RD_STRING },
    /* 17: RP */    { 2, RD_NAME, RD_NAME },
    /* 18: AFSDB */ { 2, RD_INT16, RD_NAME },
    /* 19: X25 */   { 1, RD_STRING },
    /* 20: ISDN */  { 2, RD_STRING, RD_STRING },
    /* 21: RT */    { 2, RD_INT16, RD_STRING },
    /* 22: NSAP */  { 1, RD_STRING },
    /* 23: NSAP-PTR */ {0}, /* not supported by OpenDNSSEC */
    /* 24: SIG */   { 0 }, /* not supported by OpenDNSSEC */
    /* 25: KEY */   { 0 }, /* not supported by OpenDNSSEC */
    /* 26: PX */    { 3, RD_INT16, RD_NAME, RD_NAME },
    /* 27: GPOS */  { 0 }, /* not supported by OpenDNSSEC */
    /* 28: AAAA */  { 1, RD_AAAA },
    /* 29: LOC */   { 1, RD_LOC },
    /* 30: NXT */   { 0 }, /* not supported by OpenDNSSEC */
    /* 31: EID */   { 0 }, /* not supported by OpenDNSSEC */
    /* 32: NIMLOC */{ 0 }, /* not supported by OpenDNSSEC */
    /* 33: SRV */   { 4, RD_INT16, RD_INT16, RD_INT16, RD_NAME },
    /* 34: ATMA */  { 0 },  /* not supported by OpenDNSSEC */
    /* 35: NAPTR */ { 6, RD_INT16, RD_INT16,
                      RD_STRING, RD_STRING, RD_STRING, RD_NAME },
    /* 36: KX */    { 2, RD_INT16, RD_NAME },
    /* 37: CERT */  { 4, RD_INT16, RD_INT16, RD_INT8, RD_BASE64 },
    /* 38: A6 */    { 0 }, /* not supported by OpenDNSSEC */
    /* 39: DNAME */ { 1, RD_NAME },
    /* 40: SINK */  { 0 }, /* not supported by OpenDNSSEC */
    /* 41: OPT */   { 0 }, /* not allowed in master file */
    /* 42: APL */   { 0 }, /* not supported by OpenDNSSEC */
    /* 43: DS */    { 4, RD_INT16, RD_INT8, RD_INT8, RD_BASE16 },
    /* 44: SSHFP */ { 3, RD_INT8, RD_INT8, RD_BASE16 },
    /* **FIXME** 45: IPSECKEY */ {5, RD_INT8, RD_INT8, RD_INT8, RD_STRING, RD_STRING},
    /* 46: RRSIG */ { 0 }, /* OpenDNSSEC will discard this DNSSEC RR */
    /* 47: NSEC */  { 0 }, /* OpenDNSSEC will discard this DNSSEC RR */
    /* **FIXME** 48: DNSKEY */{ 4, RD_INT16, RD_INT8, RD_INT8, RD_STRING },
    /* 49: DHCID */ { 1, RD_BASE64 },
    /* 50: NSEC3 */ { 0 }, /* OpenDNSSEC will discard this DNSSEC RR */
    /* 51: NSEC3PARAM */ { 0 }, /* OpenDNSSEC will discard this DNSSEC RR */
    /* 52: Unassigned */ { 0 },
    /* 53: Unassigned */ { 0 },
    /* 54: Unassigned */ { 0 },
    /* 55: HIP */   { 0 }, /* not supported by OpenDNSSEC */
    /* 56: NINFO */ { 0 }, /* not supported by OpenDNSSEC */
    /* 57: RKEY */  { 0 }, /* not supported by OpenDNSSEC */
    /* 58: Unassigned */ { 0 },
    /* 59: Unassigned */ { 0 },
    /* 60-69: unassigned*/ {0},{0},{0},{0},{0},{0},{0},{0},{0},{0},
    /* 70-79: unassigned*/ {0},{0},{0},{0},{0},{0},{0},{0},{0},{0},
    /* 80-89: unassigned*/ {0},{0},{0},{0},{0},{0},{0},{0},{0},{0},
    /* 90-97: unassigned*/ {0},{0},{0},{0},{0},{0},{0},{0},{0},
    /* 99: SPF */   { 1, RD_STRING },
    /* 100: Used for 32769: DLV */ { 4, RD_INT16, RD_INT8, RD_INT8, RD_BASE16 }
};

static inline void encode_int32(unsigned int val, char* dest)
{
    *((unsigned int*)dest) = val;
}

static inline unsigned int decode_int32(char* src)
{
    return *((unsigned int*)src);
}

static inline void encode_int16(unsigned int val, char* dest)
{
    *((unsigned short*)dest) = val;
}

static inline unsigned int decode_int16(char* src)
{
    return *((unsigned short*)src);
}

static int parse_ttl(char* ttl)
{
    int seconds = 0;

    while (*ttl) {
        while (isspace((unsigned char)*ttl))
            ttl++;
        int val = atoi(ttl);
        while (isdigit((unsigned char)*ttl))
            ttl++;
        switch (tolower(*ttl)) {
            case 'm': val *= 60; break;
            case 'h': val *= 3600; break;
            case 'd': val *= 86400; break;
            case 'w': val *= 86400*7; break;
            default:
                break;
        }
        seconds += val;
        ttl++;
    }
    return seconds;
}

static void encode_string(char** _src, char** _dest, bool domain_name)
{
    char* src = *_src;
    char* dest = *_dest;
    int len = 1; /* first byte is length */
    bool quoted = false;
    while (*src && (quoted || !isspace((unsigned char)*src))) {
        switch (*src) {
            case '\"':
                quoted = !quoted;
                dest[len++] = *src;
                break;

            case '\\':
                if (isdigit((unsigned char)src[1])) {
                    dest[len++] =
                        (src[1] - 48) * 100 +
                        (src[2] - 48) * 10 +
                        (src[3] - 48);
                    src += 3;
                }
                else {
                    dest[len++] = src[1];
                    src++;
                }
                break;

            case '.':
                if (domain_name) {
                    dest[0] = len - 1;
                    dest += len;
                    len = 1;
                }
                else
                    dest[len++] = *src;
                break;

            default:
                if (domain_name)
                    dest[len++] = tolower((unsigned char)*src);
                else
                    dest[len++] = *src;
                break;
        }
        src++;
    }

    dest[0] = len - 1;
    dest += len;

    /* null terminate label serie
    if (domain_name)
        *dest++ = 0; */

    if (len > 255) {
        printf("String '%s' is too long! (%d bytes)\n", src, len);
        exit(-1);
    }

    *_src = src;
    *_dest = dest;
}

static void decode_string(char** _src, char** _dest, bool domain_name)
{
    char* src = *_src;
    char* dest = *_dest;
    
    do {
        bool quoted = false;
        int len = (unsigned char)*src++;
        
        int i;
        for (i=0; i<len; i++) {
            if (*src == '\"')
                quoted = !quoted;

            if (isprint((unsigned char)*src))
                *dest++ = *src;
            else
                dest += sprintf(dest, "\\%03d", *src);
            src++;
        }
        if (domain_name)
            *dest++ = '.';
    } while (domain_name && *src);

    if (domain_name)
        src++;

    *_src = src;
    *_dest = dest;
}

static void* encode_owner(char* name, void* dest)
{
    char tmpname[MAX_NAME_LEN];
    char labelpos[MAX_NAME_LEN];
    char* tmpptr = tmpname;
    encode_string(&name, &tmpptr, true);

    /* iterate through string and store the position of each label */
    int count = 0;
    int pos = 0;
    do {
        labelpos[count++] = pos;
        pos += tmpname[pos] + 1;
    } while (tmpname[pos]);

    /* now store the name backwards */
    short* dptr = dest;
    int i;
    for (i=count-1; i >= 0; i--) {
        char* sptr = tmpname + labelpos[i];
        int len = *sptr++;
        while (len--)
            *dptr++ = *sptr++;
        *dptr++ = -2;
    }
    *dptr++ = -1;

    return dptr;
}

static void decode_owner(char** _rr, char** _dest)
{
    short* rr = (short*)*_rr;
    char* dest = *_dest;

    int len = 0;
    while (rr[len] != -1)
        len++;

    /* reverse name */
    short* delim = rr + len - 1;

    while (1) {
        short* end = delim;

        /* find start of name segment */
        while (delim > rr && (delim[-1] != -2))
            delim--;

        short* p = delim;
        
        /* copy segment */
        while (p < end)
            *dest++ = *p++;
        *dest++ = '.';

        if (delim == rr)
            break;

        delim--;
    }
    *_rr += (len + 1) * sizeof(short);
    *_dest = dest;
}

void encode_int(char** src, char** dest, int type)
{
    int val = atoi(*src);
    switch (type) {
        case RD_INT8:
            **dest = val;
            (*dest)++;
            break;

        case RD_INT16:
            encode_int16(val, *dest);
            (*dest) += 2;
            break;

        case RD_INT32:
            encode_int32(val, *dest);
            (*dest) += 4;
            break;
    }

    while (**src && !isspace((unsigned)**src))
        (*src)++;
}

static void* encode_rdata(int type, char* rdata, char* dest)
{
    const char* format = format_list[type];
    int pcount = format[0];

    int i;
    for (i=1; i <= pcount; i++) {
        switch (format[i]) {
            case RD_NAME:
                encode_string(&rdata, &dest, true);
                break;

            case RD_STRING:
                encode_string(&rdata, &dest, false);
                break;

            case RD_A:
                inet_pton(AF_INET, rdata, dest);
                while (isspace((unsigned char)*rdata))
                    rdata++;
                dest += 4;
                break;

            case RD_AAAA:
                inet_pton(AF_INET6, rdata, dest);
                while (isspace((unsigned char)*rdata))
                    rdata++;
                dest += 16;
                break;
                
            case RD_INT8:
            case RD_INT16:
            case RD_INT32:
                encode_int(&rdata, &dest, format[i]);
                break;

            default:
                printf("Error! Unsupported type %d.\n", format[i]);
                exit(-1);
                break;
        }
        while (isspace((unsigned char)*rdata))
            rdata++;
    }

    return dest;
}

static void decode_rdata(int type, char* rdata, char* dest)
{
    const char* format = format_list[type];
    int pcount = format[0];

    int i;
    for (i=1; i <= pcount; i++) {
        switch (format[i]) {
            case RD_NAME:
                decode_string(&rdata, &dest, true);
                break;

            case RD_STRING:
                decode_string(&rdata, &dest, false);
                break;

            case RD_A:
                inet_ntop(AF_INET, rdata, dest, INET_ADDRSTRLEN);
                rdata += 4;
                while (*dest)
                    dest++;
                break;

            case RD_AAAA:
                inet_ntop(AF_INET6, rdata, dest, INET6_ADDRSTRLEN);
                rdata += 16;
                while (*dest)
                    dest++;
                break;
                
            case RD_INT8:
                dest += sprintf(dest, "%d", *rdata);
                rdata++;
                break;

            case RD_INT16:
                dest += sprintf(dest, "%d", decode_int16(rdata));
                rdata += 2;
                break;

            case RD_INT32:
                dest += sprintf(dest, "%d", decode_int32(rdata));
                rdata += 4;
                break;
        }
        if (i<pcount)
            *dest++ = ' ';
    }
    *dest = 0;
}

int encode_rr(char* name, int type, int class, char* ttl, char* rdata,
               char* dest)
{
    char* tmp;

    int seconds = parse_ttl(ttl);
    encode_int32(seconds, dest);

    tmp = dest + 4;
    char* ptr = encode_owner(name, (short*)(dest+4));

    encode_int16(type, ptr);
    ptr += 2;

    encode_int16(class, ptr);
    ptr += 2;

    tmp = ptr;
    ptr = encode_rdata(type, rdata, ptr);

    return ptr - dest;
}

void decode_rr(char* src, char* dest)
{
    int ttl = decode_int32(src);
    src += 4;

    decode_owner(&src, &dest);
    *dest++ = ' ';

    dest += sprintf(dest, "%d ", ttl);

    int type = decode_int16(src);
    src += 2;
    int class = decode_int16(src);
    src += 2;

    if (class > 0 && class < NUM_CLASSES)
        dest += sprintf(dest, "%s ", classname[class]);
    else
        dest += sprintf(dest, "CLASS%d ", type);

    if (type == 32769) /* special case for DLV */
        type = 100;
    if (type > 0 && type < NUM_TYPES)
        dest += sprintf(dest, "%s ", typename[type]);
    else
        dest += sprintf(dest, "TYPE%d ", type);
        
    decode_rdata(type, src, dest);
}

void hexdump(void* ptr, int count)
{
    const int DUMP_LINE_LEN = 16;
    unsigned char* buffer = ptr;

    int l;
    for (l=0; l<=count/DUMP_LINE_LEN; l++) {
        int start = l*DUMP_LINE_LEN;
        int end = start + DUMP_LINE_LEN;
        if ( end > count )
            end = count;

        if ( start == end )
            break;

        printf("%p: ", ptr + l*DUMP_LINE_LEN);

        /* print hex code */
        int i;
        for (i=start; i<start+DUMP_LINE_LEN; i++ ) {
            if ( i < end )
                printf("%02x ", buffer[i] );
            else
                printf("   ");
        }

        /* print characters */
        for (i=start; i<end; i++ ) {
            if ( isprint( buffer[i] ) )
                printf("%c", buffer[i] );
            else
                printf(".");
        }
        printf("\n");
    }
}


int main(void)
{
    char buf1[MAX_LINE_LEN];
    char buf2[MAX_LINE_LEN];

    memset(buf1, 0x55, sizeof(buf1));
    int len = encode_rr("www.example.com.", 99, 1, "2d1h", "\"quoted string\"", buf1);
    hexdump(buf1, len);

    memset(buf2, 0xaa, sizeof(buf2));
    decode_rr(buf1, buf2);
    hexdump(buf2, strlen(buf2)+1);
    return 0;
}
